<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Tidy Transcriptomics • rladiestunis2021tidytranscriptomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Tidy Transcriptomics">
<meta property="og:description" content="rladiestunis2021tidytranscriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rladiestunis2021tidytranscriptomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.8.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/background.html">Background</a>
</li>
<li>
  <a href="../articles/tidytranscriptomics.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/rladiestunis2021_tidytranscriptomics">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/repository/docker/stemangiola/rladiestunis2021_tidytranscriptomics">
    <span class="fab fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tidytranscriptomics_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to Tidy Transcriptomics</h1>
                        <h4 class="author">Maria Doyle, Peter MacCallum Cancer Centre<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 class="author">Stefano Mangiola, Walter and Eliza Hall Institute<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/rladiestunis2021_tidytranscriptomics/blob/master/vignettes/tidytranscriptomics.Rmd"><code>vignettes/tidytranscriptomics.Rmd</code></a></small>
      <div class="hidden name"><code>tidytranscriptomics.Rmd</code></div>

    </div>

    
    
<div id="workshop-introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-introduction" class="anchor"></a>Workshop introduction</h1>
<p><a href="https://stemangiola.github.io/tidybulk/">tidybulk</a>, <a href="https://stemangiola.github.io/tidySummarizedExperiment/">tidySummarizedExperiment</a> and <a href="https://stemangiola.github.io/tidySingleCellExperiment/">tidySingleCellExperiment</a> are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data representation and analysis. The roadmap for the development of the tidytranscriptomics packages is shown in the figure below.</p>
<p><img src="../inst/vignettes/roadmap.png" width="600px"></p>
</div>
<div id="part-1-bulk-rna-seq-with-tidysummarizedexperiment-and-tidybulk" class="section level1">
<h1 class="hasAnchor">
<a href="#part-1-bulk-rna-seq-with-tidysummarizedexperiment-and-tidybulk" class="anchor"></a>Part 1 Bulk RNA-seq with tidySummarizedExperiment and tidybulk</h1>
<p><img src="../inst/vignettes/tidybulk_logo.png" width="100px"></p>
<div id="acknowledgements" class="section level3">
<h3 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h3>
<p>Some of the material in Part 1 was adapted from an R for RNA sequencing workshop first run <a href="http://combine-australia.github.io/2016-05-11-RNAseq/">here</a>. Use of the airway dataset was inspired by the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2 vignette</a>.</p>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA sequencing to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>
<p>There are many steps involved in analysing an RNA sequencing dataset. Sequenced reads are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today we will be starting from the count data and showing how differential expression analysis can be performed in a friendly way using the Bioconductor package, tidybulk.</p>
<p>First, let’s load all the packages we will need to analyse the data.</p>
<p><em>Note: you should load the <em>tidybulk</em> and <em>tidySummarizedExperiment</em> libraries after the tidyverse core packages for best integration.</em></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dataset</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># tidyverse core packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org">stringr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>

<span class="co"># tidyverse-friendly packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggobi.github.io/ggally/">GGally</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org">tidyHeatmap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidybulk</span><span class="op">)</span>
<span class="co"># library(tidySummarizedExperiment) # we'll load this below to show what it can do</span></code></pre></div>
<p>Plot settings. Set the colours and theme we will use for our plots.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Use colourblind-friendly colours</span>
<span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html">dittoColors</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Set theme</span>
<span class="va">custom_theme</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
        panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span><span class="op">)</span>,
        panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,
        panel.grid.minor <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,
        text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">10</span>, r <span class="op">=</span> <span class="fl">10</span>, b <span class="op">=</span> <span class="fl">10</span>, l <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">30</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
      <span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<div id="tidysummarizedexperiment" class="section level3">
<h3 class="hasAnchor">
<a href="#tidysummarizedexperiment" class="anchor"></a>tidySummarizedExperiment</h3>
<p>Here we will perform our analysis using the data from the <em>airway</em> package. The airway data comes from the paper by <span class="citation">(Himes et al. 2014)</span>; and it includes 8 samples from human airway smooth muscle cells, from 4 cell lines. For each cell line treated (with dexamethasone) and untreated (negative control) a sample has undergone RNA sequencing and gene counts have been generated.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load airway RNA sequencing data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">airway</span>
<span class="co">#&gt; class: RangedSummarizedExperiment </span>
<span class="co">#&gt; dim: 64102 8 </span>
<span class="co">#&gt; metadata(1): ''</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(64102): ENSG00000000003 ENSG00000000005 ... LRG_98 LRG_99</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span>
<span class="co">#&gt; colData names(9): SampleName cell ... Sample BioSample</span></code></pre></div>
<p>The data in the airway package is a Bioconductor <em>SummarizedExperiment</em> object. For more information see <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">here</a>.</p>
<p>The <code>tidySummarizedExperiment</code> package enables any <em>SummarizedExperiment</em> object to be displayed and manipulated according to tidy data principles, without affecting any <em>SummarizedExperiment</em> behaviour.</p>
<p>If we load the <em>tidySummarizedExperiment</em> package and then view the airway data it now displays as a tibble. A tibble is the tidyverse table format.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load tidySummarizedExperiment package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidySummarizedExperiment</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">airway</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 512,816 x 13</span>
<span class="co">#&gt; [90m# Transcripts=64102 | Samples=8 | Assays=counts[39m</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 40 more rows, and 3 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   coordinate &lt;list&gt;</span></code></pre></div>
<p>Now we can more easily see the data. The airway object contains information about genes and samples, the first column has the Ensembl gene identifier, the second column has the sample identifier and the third column has the gene transcription abundance. The abundance is the number of reads aligning to the gene in each experimental sample. The remaining columns include sample information. The dex column tells us whether the samples are treated or untreated and the cell column tells us what cell line they are from.</p>
<p>We can still interact with the tidy SummarizedExperiment object using commands for SummarizedExperiment objects.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assays</span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="co">#&gt; List of length 1</span>
<span class="co">#&gt; names(1): counts</span></code></pre></div>
</div>
<div id="tidyverse-commands" class="section level3">
<h3 class="hasAnchor">
<a href="#tidyverse-commands" class="anchor"></a>Tidyverse commands</h3>
<p>And now we can also use tidyverse commands, such as <code>filter</code>, <code>select</code> and <code>mutate</code> to explore the tidy SummarizedExperiment object. Some examples are shown below and more can be seen at the tidySummarizedExperiment website <a href="https://stemangiola.github.io/tidySummarizedExperiment/articles/introduction.html#tidyverse-commands-1">here</a>. We can also use the tidyverse pipe <code><a href="https://purrr.tidyverse.org/reference/pipe.html">%&gt;%</a></code>. This ‘pipes’ the output from the command on the left into the command on the right/below. Using the pipe is not essential but it reduces the amount of code we need to write when we have multiple steps. It also can make the steps clearer and easier to see. For more details on the pipe see <a href="https://r4ds.had.co.nz/pipes.html">here</a>.</p>
<p>We can use <code>filter</code> to choose rows, for example, to see just the rows for the treated samples.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">airway</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">filter</a></span><span class="op">(</span><span class="va">dex</span> <span class="op">==</span> <span class="st">"trt"</span><span class="op">)</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 256,408 x 13</span>
<span class="co">#&gt; [90m# Transcripts=64102 | Samples=4 | Assays=counts[39m</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    448 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    515 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    211 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     55 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3679 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1062 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    380 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    236 GSM1275863 N613… trt   untrt SRR1…       126 SRX384346 </span>
<span class="co">#&gt; # … with 40 more rows, and 3 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   coordinate &lt;list&gt;</span></code></pre></div>
<p>We can use <code>select</code> to choose columns, for example, to see the sample, cell line and treatment columns.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">airway</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">cell</span>, <span class="va">dex</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 512,816 x 3</span>
<span class="co">#&gt;    sample     cell   dex  </span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;fct&gt;  &lt;fct&gt;</span>
<span class="co">#&gt;  1 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  2 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  3 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  4 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  5 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  6 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  7 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  8 SRR1039508 N61311 untrt</span>
<span class="co">#&gt;  9 SRR1039508 N61311 untrt</span>
<span class="co">#&gt; 10 SRR1039508 N61311 untrt</span>
<span class="co">#&gt; # … with 512,806 more rows</span></code></pre></div>
<p>We can use <code>mutate</code> to create a column. For example, we could create a new sample_name column that contains shorter sample names. We can remove the SRR1039 prefix that’s present in all of the samples, as shorter names can fit better in some of the plots we will create. We can use <code>mutate</code> together with <code>str_replace</code> to remove the SRR1039 string from the sample column.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>sample_name<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"SRR1039"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># select columns to view    </span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">sample_name</span><span class="op">)</span>
<span class="co">#&gt; tidySummarizedExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 512,816 x 2</span>
<span class="co">#&gt;    sample     sample_name</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;      </span>
<span class="co">#&gt;  1 SRR1039508 508        </span>
<span class="co">#&gt;  2 SRR1039508 508        </span>
<span class="co">#&gt;  3 SRR1039508 508        </span>
<span class="co">#&gt;  4 SRR1039508 508        </span>
<span class="co">#&gt;  5 SRR1039508 508        </span>
<span class="co">#&gt;  6 SRR1039508 508        </span>
<span class="co">#&gt;  7 SRR1039508 508        </span>
<span class="co">#&gt;  8 SRR1039508 508        </span>
<span class="co">#&gt;  9 SRR1039508 508        </span>
<span class="co">#&gt; 10 SRR1039508 508        </span>
<span class="co">#&gt; # … with 512,806 more rows</span></code></pre></div>
</div>
</div>
<div id="setting-up-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-data" class="anchor"></a>Setting up the data</h2>
<p>We’ll set up the airway data for our RNA-seq analysis. We’ll create the column with shorter sample names and a column with gene symbols. We can get the gene symbols for these Ensembl gene ids using the Bioconductor annotation package for human, <code>org.Hs.eg.db</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># setup data workflow</span>
<span class="va">counts</span> <span class="op">&lt;-</span>
  <span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"SRR1039"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">mapIds</a></span><span class="op">(</span><span class="fu">org.Hs.eg.db</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/org.Hs.eg.db/man/org.Hs.egBASE.html">org.Hs.eg.db</a></span>,
    keys <span class="op">=</span> <span class="va">feature</span>,
    keytype <span class="op">=</span> <span class="st">"ENSEMBL"</span>,
    column <span class="op">=</span> <span class="st">"SYMBOL"</span>,
    multiVals <span class="op">=</span> <span class="st">"first"</span>
  <span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span>
<span class="co"># take a look</span>
<span class="va">counts</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 512,816 x 15</span>
<span class="co">#&gt; [90m# Transcripts=64102 | Samples=8 | Assays=counts[39m</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 40 more rows, and 5 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   sample_name &lt;chr&gt;, symbol &lt;chr&gt;, coordinate &lt;list&gt;</span></code></pre></div>
</div>
<div id="filtering-lowly-transcribed-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-lowly-transcribed-genes" class="anchor"></a>Filtering lowly transcribed genes</h2>
<p>Genes with very low counts across all libraries provide little evidence for differential expression and they can interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.</p>
<p>We can perform the filtering using tidybulk <code>keep_abundant</code> or <code>identify_abundant</code>. These functions can use the <em>edgeR</em> <code>filterByExpr</code> function described in <span class="citation">(Law et al. 2016)</span> to automatically identify the genes with adequate abundance for differential expression testing. By default, this will keep genes with ~10 counts in a minimum number of samples, the number of the samples in the smallest group. In this dataset the smallest group size is four (as we have four dex-treated samples versus four untreated). Alternatively, we could use <code>identify_abundant</code> to identify which genes are abundant or not (TRUE/FALSE), rather than just keeping the abundant ones.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Filtering counts</span>
<span class="va">counts_filtered</span> <span class="op">&lt;-</span> <span class="va">counts</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_filtered</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 127,408 x 16</span>
<span class="co">#&gt; [90m# Transcripts=15926 | Samples=8 | Assays=counts[39m</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    172 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…   2112 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 40 more rows, and 6 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;,</span>
<span class="co">#&gt; #   sample_name &lt;chr&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, coordinate &lt;list&gt;</span></code></pre></div>
<p>After running <code>keep_abundant</code> we have a column called <code>.abundant</code> containing TRUE (<code>identify_abundant</code> would have TRUE/FALSE).</p>
</div>
<div id="scaling-counts-to-normalise" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-counts-to-normalise" class="anchor"></a>Scaling counts to normalise</h2>
<p>Scaling of counts, normalisation, is performed to eliminate uninteresting differences between samples due to sequencing depth or composition. A more detailed explanation can be found <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">here</a>. In the tidybulk package the function <code>scale_abundance</code> generates scaled counts, with scaling factors calculated on abundant (filtered) transcripts and applied to all transcripts. We can choose from different normalisation methods. Here we will use the default, edgeR’s trimmed mean of M values (TMM), <span class="citation">(Robinson and Oshlack 2010)</span>. TMM normalisation (and most scaling normalisation methods) scale relative to one sample.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Scaling counts</span>
<span class="va">counts_scaled</span> <span class="op">&lt;-</span> <span class="va">counts_filtered</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># take a look</span>
<span class="va">counts_scaled</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 127,408 x 19</span>
<span class="co">#&gt; [90m# Transcripts=15926 | Samples=8 | Assays=counts, counts_scaled[39m</span>
<span class="co">#&gt;    feature sample counts counts_scaled SampleName cell  dex   albut Run  </span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt;         &lt;dbl&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;</span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679         961.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  2 ENSG00… SRR10…    467         661.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  3 ENSG00… SRR10…    260         368.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  4 ENSG00… SRR10…     60          84.9 GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  5 ENSG00… SRR10…   3251        4601.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  6 ENSG00… SRR10…   1433        2028.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  7 ENSG00… SRR10…    519         734.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  8 ENSG00… SRR10…    394         558.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  9 ENSG00… SRR10…    172         243.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; 10 ENSG00… SRR10…   2112        2989.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; # … with 40 more rows, and 10 more variables: avgLength &lt;int&gt;,</span>
<span class="co">#&gt; #   Experiment &lt;fct&gt;, Sample &lt;fct&gt;, BioSample &lt;fct&gt;, sample_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, symbol &lt;chr&gt;, .abundant &lt;lgl&gt;,</span>
<span class="co">#&gt; #   coordinate &lt;list&gt;</span></code></pre></div>
<p>After we run <code>scale_abundance</code> we should see some columns have been added at the end. The <code>counts_scaled</code> column contains the scaled counts.</p>
<p>We can visualise the difference of abundance densities before and after scaling. As tidybulk output is compatible with tidyverse, we can simply pipe it into standard tidyverse functions such as <code>filter</code>, <code>pivot_longer</code> and <code>ggplot</code>. We can also take advantage of ggplot’s <code>facet_wrap</code> to easily create multiple plots.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scaled</span> <span class="op">%&gt;%</span>

  <span class="co"># Reshaping</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tidyr-methods.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"counts_scaled"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"source"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plotting</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="va">sample_name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span>
<span class="co">#&gt; tidySummarizedExperiment says: A data frame is returned for independent data analysis.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-15-1.png" width="70%"></p>
<p>In this dataset the distributions of the counts are not very different to each other before scaling but scaling does make the distributions more similar. If we saw a sample with a very different distribution we may need to investigate it.</p>
<p>As tidybulk smoothly integrates with ggplot2 and other tidyverse packages it can save on typing and make plots easier to generate. Compare the code for creating density plots with tidybulk versus standard base R below (standard code adapted from <span class="citation">(Law et al. 2016)</span>).</p>
<p><strong>tidybulk</strong></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># tidybulk</span>
<span class="va">airway</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tidyr-methods.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"counts"</span>, <span class="st">"counts_scaled"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"source"</span>, values_to <span class="op">=</span> <span class="st">"abundance"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">source</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><strong>base R using edgeR</strong></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="co"># Prepare data set</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">nsamples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>

<span class="co"># Setup graphics</span>
<span class="va">col</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="va">nsamples</span>, <span class="st">"Paired"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Plot raw counts</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.26</span><span class="op">)</span>, las <span class="op">=</span> <span class="fl">2</span>, main <span class="op">=</span> <span class="st">""</span>, xlab <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span>main <span class="op">=</span> <span class="st">"Counts"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsamples</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">den</span><span class="op">$</span><span class="va">x</span>, <span class="va">den</span><span class="op">$</span><span class="va">y</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend <span class="op">=</span> <span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Run</span>, text.col <span class="op">=</span> <span class="va">col</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span>

<span class="co"># Plot scaled counts</span>
<span class="va">dgList_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">lcpm_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">dgList_norm</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">lcpm_n</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.26</span><span class="op">)</span>, las <span class="op">=</span> <span class="fl">2</span>, main <span class="op">=</span> <span class="st">""</span>, xlab <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"Counts scaled"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">nsamples</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span><span class="op">(</span><span class="va">lcpm_n</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">den</span><span class="op">$</span><span class="va">x</span>, <span class="va">den</span><span class="op">$</span><span class="va">y</span>, col <span class="op">=</span> <span class="va">col</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend <span class="op">=</span> <span class="va">dgList_norm</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Run</span>, text.col <span class="op">=</span> <span class="va">col</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></code></pre></div>
</div>
<div id="exploratory-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#exploratory-analyses" class="anchor"></a>Exploratory analyses</h2>
<div id="dimensionality-reduction" class="section level3">
<h3 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality reduction</h3>
<p>By far, one of the most important plots we make when we analyse RNA sequencing data are principal-component analysis (PCA) or multi-dimensional scaling (MDS) plots. We reduce the dimensions of the data to identify the greatest sources of variation in the data. A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. We can use the <code>reduce_dimensions</code> function to calculate the dimensions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get principal components</span>
<span class="va">counts_scal_PCA</span> <span class="op">&lt;-</span>
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span>
<span class="co">#&gt; Fraction of variance explained by the selected principal components</span>
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   `Fraction of variance`    PC</span>
<span class="co">#&gt;                    &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1                  0.323     1</span>
<span class="co">#&gt; 2                  0.305     2</span>
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span></code></pre></div>
<p>This joins the result to the counts object.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Take a look</span>
<span class="va">counts_scal_PCA</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 127,408 x 22</span>
<span class="co">#&gt; [90m# Transcripts=15926 | Samples=8 | Assays=counts, counts_scaled[39m</span>
<span class="co">#&gt;    feature sample counts counts_scaled SampleName cell  dex   albut Run  </span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt;         &lt;dbl&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;</span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679         961.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  2 ENSG00… SRR10…    467         661.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  3 ENSG00… SRR10…    260         368.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  4 ENSG00… SRR10…     60          84.9 GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  5 ENSG00… SRR10…   3251        4601.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  6 ENSG00… SRR10…   1433        2028.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  7 ENSG00… SRR10…    519         734.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  8 ENSG00… SRR10…    394         558.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  9 ENSG00… SRR10…    172         243.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; 10 ENSG00… SRR10…   2112        2989.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; # … with 40 more rows, and 13 more variables: avgLength &lt;int&gt;,</span>
<span class="co">#&gt; #   Experiment &lt;fct&gt;, Sample &lt;fct&gt;, BioSample &lt;fct&gt;, sample_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, sample.x &lt;chr&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, coordinate &lt;list&gt;</span></code></pre></div>
<p>For plotting, we can select just the sample-wise information with <code>pivot_sample</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># take a look</span>
<span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 8 x 16</span>
<span class="co">#&gt;   sample    SampleName cell    dex   albut Run      avgLength Experiment Sample </span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;fct&gt;      &lt;fct&gt;   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;        &lt;int&gt; &lt;fct&gt;      &lt;fct&gt;  </span>
<span class="co">#&gt; 1 SRR10395… GSM1275862 N61311  untrt untrt SRR1039…       126 SRX384345  SRS508…</span>
<span class="co">#&gt; 2 SRR10395… GSM1275863 N61311  trt   untrt SRR1039…       126 SRX384346  SRS508…</span>
<span class="co">#&gt; 3 SRR10395… GSM1275866 N052611 untrt untrt SRR1039…       126 SRX384349  SRS508…</span>
<span class="co">#&gt; 4 SRR10395… GSM1275867 N052611 trt   untrt SRR1039…        87 SRX384350  SRS508…</span>
<span class="co">#&gt; 5 SRR10395… GSM1275870 N080611 untrt untrt SRR1039…       120 SRX384353  SRS508…</span>
<span class="co">#&gt; 6 SRR10395… GSM1275871 N080611 trt   untrt SRR1039…       126 SRX384354  SRS508…</span>
<span class="co">#&gt; 7 SRR10395… GSM1275874 N061011 untrt untrt SRR1039…       101 SRX384357  SRS508…</span>
<span class="co">#&gt; 8 SRR10395… GSM1275875 N061011 trt   untrt SRR1039…        98 SRX384358  SRS508…</span>
<span class="co">#&gt; # … with 7 more variables: BioSample &lt;fct&gt;, sample_name &lt;chr&gt;, TMM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   multiplier &lt;dbl&gt;, sample.x &lt;chr&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;</span></code></pre></div>
<p>We can now plot the reduced dimensions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># PCA plot</span>
<span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, colour <span class="op">=</span> <span class="va">dex</span>, shape <span class="op">=</span> <span class="va">cell</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">sample_name</span><span class="op">)</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-21-1.png" width="70%"></p>
<p>The samples separate by treatment on PC1 which is what we hope to see. PC2 separates the N080611 cell line from the other samples, indicating a greater difference between that cell line and the others.</p>
</div>
<div id="hierarchical-clustering-with-heatmaps" class="section level3">
<h3 class="hasAnchor">
<a href="#hierarchical-clustering-with-heatmaps" class="anchor"></a>Hierarchical clustering with heatmaps</h3>
<p>An alternative to principal component analysis for examining relationships between samples is using hierarchical clustering. Heatmaps are a nice visualisation to examine hierarchical clustering of your samples. tidybulk has a simple function we can use, <code>keep_variable</code>, to extract the most variable genes which we can then plot with tidyHeatmap.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>

  <span class="co"># extract 500 most variable genes</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_variable-methods.html">keep_variable</a></span><span class="op">(</span>.abundance <span class="op">=</span> <span class="va">counts_scaled</span>, top <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tibble-methods.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># create heatmap</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap-methods.html">heatmap</a></span><span class="op">(</span>
    .column <span class="op">=</span> <span class="va">sample_name</span>,
    .row <span class="op">=</span> <span class="va">feature</span>,
    .value <span class="op">=</span> <span class="va">counts_scaled</span>,
    transform <span class="op">=</span> <span class="va">log1p</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/add_tile-methods.html">add_tile</a></span><span class="op">(</span><span class="va">dex</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/add_tile-methods.html">add_tile</a></span><span class="op">(</span><span class="va">cell</span><span class="op">)</span>
<span class="co">#&gt; Getting the 500 most variable genes</span>
<span class="co">#&gt; Warning in add_annotation(., !!.column, type = "tile", palette_discrete</span>
<span class="co">#&gt; = .data@data %&gt;% : tidyHeatmap says: nested/list column are present in your data</span>
<span class="co">#&gt; frame and have been dropped as their unicity cannot be identified by dplyr.</span>

<span class="co">#&gt; Warning in add_annotation(., !!.column, type = "tile", palette_discrete</span>
<span class="co">#&gt; = .data@data %&gt;% : tidyHeatmap says: nested/list column are present in your data</span>
<span class="co">#&gt; frame and have been dropped as their unicity cannot be identified by dplyr.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-22-1.png" width="70%"></p>
<p>In the heatmap we can see the samples cluster into two groups, treated and untreated, for three of the cell lines, and the cell line (N080611) again is further away from the others.</p>
<p>Tidybulk enables a simplified way of generating a clustered heatmap of variable genes. Compare the code below for tidybulk versus a base R method.</p>
<p><strong>base R using edgeR</strong></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">dgList</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span>
<span class="va">select_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">var_genes</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span>
<span class="va">highly_variable_lcpm</span> <span class="op">&lt;-</span> <span class="va">logcounts</span><span class="op">[</span><span class="va">select_var</span>, <span class="op">]</span>
<span class="va">colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#fefada"</span><span class="op">)</span>
<span class="va">col.group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"grey"</span><span class="op">)</span><span class="op">[</span><span class="va">group</span><span class="op">]</span>
<span class="fu">gplots</span><span class="fu">::</span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">highly_variable_lcpm</span>, col <span class="op">=</span> <span class="va">colours</span>, trace <span class="op">=</span> <span class="st">"none"</span>, ColSideColors <span class="op">=</span> <span class="va">col.group</span>, scale <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h2>
<p><em>tidybulk</em> integrates several popular methods for differential transcript abundance testing: the edgeR quasi-likelihood <span class="citation">(Chen, Lun, and Smyth 2016)</span> (tidybulk default method), edgeR likelihood ratio <span class="citation">(McCarthy, Chen, and Smyth 2012)</span>, limma-voom <span class="citation">(Law et al. 2014)</span> and DESeq2 <span class="citation">(Love, Huber, and Anders 2014)</span>. A common question researchers have is which method to choose. With tidybulk we can easily run multiple methods and compare.</p>
<p>We give <code>test_differential_abundance</code> our tidybulk counts object and a formula, specifying the column that contains our groups to be compared. If all our samples were from the same cell line, and there were no additional factors contributing variance such as batch differences, we could use the formula <code>~ dex</code>. However, each treated and untreated sample is from a different cell line so we add the cell line as an additional factor <code>~ dex + cell</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_all</span> <span class="op">&lt;-</span>

  <span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>

  <span class="co"># edgeR QLT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"edgeR_quasi_likelihood"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerQLT_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># edgeR LRT</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"edgeR_likelihood_ratio"</span>,
    prefix <span class="op">=</span> <span class="st">"edgerLR_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># limma-voom</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"limma_voom"</span>,
    prefix <span class="op">=</span> <span class="st">"voom_"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># DESeq2</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    <span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span>,
    method <span class="op">=</span> <span class="st">"deseq2"</span>,
    prefix <span class="op">=</span> <span class="st">"deseq2_"</span>
  <span class="op">)</span>

<span class="co"># take a look</span>

<span class="va">de_all</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 127,408 x 44</span>
<span class="co">#&gt; [90m# Transcripts=15926 | Samples=8 | Assays=counts, counts_scaled[39m</span>
<span class="co">#&gt;    feature sample counts counts_scaled SampleName cell  dex   albut Run  </span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt;         &lt;dbl&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;</span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679         961.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  2 ENSG00… SRR10…    467         661.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  3 ENSG00… SRR10…    260         368.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  4 ENSG00… SRR10…     60          84.9 GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  5 ENSG00… SRR10…   3251        4601.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  6 ENSG00… SRR10…   1433        2028.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  7 ENSG00… SRR10…    519         734.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  8 ENSG00… SRR10…    394         558.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt;  9 ENSG00… SRR10…    172         243.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; 10 ENSG00… SRR10…   2112        2989.  GSM1275862 N613… untrt untrt SRR1…</span>
<span class="co">#&gt; # … with 40 more rows, and 35 more variables: avgLength &lt;int&gt;,</span>
<span class="co">#&gt; #   Experiment &lt;fct&gt;, Sample &lt;fct&gt;, BioSample &lt;fct&gt;, sample_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, sample.x &lt;chr&gt;, PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   symbol &lt;chr&gt;, .abundant &lt;lgl&gt;, edgerQLT_logFC &lt;dbl&gt;, edgerQLT_logCPM &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerQLT_F &lt;dbl&gt;, edgerQLT_PValue &lt;dbl&gt;, edgerQLT_FDR &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_logFC &lt;dbl&gt;, edgerLR_logCPM &lt;dbl&gt;, edgerLR_LR &lt;dbl&gt;,</span>
<span class="co">#&gt; #   edgerLR_PValue &lt;dbl&gt;, edgerLR_FDR &lt;dbl&gt;, voom_logFC &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_AveExpr &lt;dbl&gt;, voom_t &lt;dbl&gt;, voom_P.Value &lt;dbl&gt;, voom_adj.P.Val &lt;dbl&gt;,</span>
<span class="co">#&gt; #   voom_B &lt;dbl&gt;, deseq2_baseMean &lt;dbl&gt;, deseq2_log2FoldChange &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_lfcSE &lt;dbl&gt;, deseq2_stat &lt;dbl&gt;, deseq2_pvalue &lt;dbl&gt;,</span>
<span class="co">#&gt; #   deseq2_padj &lt;dbl&gt;, coordinate &lt;list&gt;</span></code></pre></div>
<p>This outputs the columns from each method such as log-fold change (logFC), false-discovery rate (FDR) and probability value (p-value). logFC is log2(treated/untreated).</p>
<div id="comparison-of-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#comparison-of-methods" class="anchor"></a>Comparison of methods</h3>
<p>We can visually compare the significance for all methods. We will notice that there is some difference between the methods.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de_all</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">edgerQLT_PValue</span>, <span class="va">edgerLR_PValue</span>, <span class="va">voom_P.Value</span>, <span class="va">deseq2_pvalue</span>, <span class="va">transcript</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
</div>
<div id="single-method" class="section level3">
<h3 class="hasAnchor">
<a href="#single-method" class="anchor"></a>Single method</h3>
<p>If we just wanted to run one differential testing method we could do that. The default method is edgeR quasi-likelihood.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_de</span> <span class="op">&lt;-</span> <span class="va">counts_scal_PCA</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span><span class="op">~</span> <span class="va">dex</span> <span class="op">+</span> <span class="va">cell</span><span class="op">)</span>
<span class="co">#&gt; =====================================</span>
<span class="co">#&gt; tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance </span>
<span class="co">#&gt; or adjust_abundance have been calculated. Therefore, it is essential to add covariates </span>
<span class="co">#&gt; such as batch effects (if applicable) in the formula.</span>
<span class="co">#&gt; =====================================</span>
<span class="co">#&gt; tidybulk says: The design column names are "(Intercept), dexuntrt, cellN061011, cellN080611, cellN61311"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR_quasi_likelihood`</span></code></pre></div>
<p>Tidybulk enables a simplified way of performing an RNA sequencing differential expression analysis (with the benefit of smoothly integrating with ggplot2 and other tidyverse packages). Compare the code for a tidybulk edgeR analysis versus standard edgeR below.</p>
<p><strong>standard edgeR</strong></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example code, no need to run</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">dex</span><span class="op">)</span>
<span class="va">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">dgList</span>, group <span class="op">=</span> <span class="va">group</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="va">dgList</span><span class="op">[</span><span class="va">keep.exprs</span>, , keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">)</span>
<span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dgList</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">cell</span><span class="op">)</span>
<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">group</span> <span class="op">+</span> <span class="va">cell</span><span class="op">)</span>
<span class="va">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">dgList</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFit</a></span><span class="op">(</span><span class="va">dgList</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="volcano-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#volcano-plots" class="anchor"></a>Volcano plots</h2>
<p>Volcano plots are a useful genome-wide plot for checking that the analysis looks good. Volcano plots enable us to visualise the significance of change (p-value) versus the fold change (logFC). Highly significant genes are towards the top of the plot. We can also colour significant genes e.g. genes with false-discovery rate &lt; 0.05. In order to decide which genes are differentially expressed, we usually take a cut-off of 0.05 on the FDR (or adjusted P value), NOT the raw p-value. This is because we are testing many genes (multiple testing), and the chances of finding differentially expressed genes is very high when you do that many tests. Hence we need to control the false discovery rate, which is the adjusted p-value column in the results table. What this means is that if 100 genes are significant at a 5% false discovery rate, we are willing to accept that 5 will be false positives.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># volcano plot, minimal</span>
<span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span>, colour <span class="op">=</span> <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-28-1.png" width="70%"></p>
<p>We’ll extract the symbols for a few top genes (by P value) to use in a more informative volcano plot, integrating some of the packages in tidyverse.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">topgenes_symbols</span> <span class="op">&lt;-</span>
  <span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">pull</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>

<span class="va">topgenes_symbols</span>
<span class="co">#&gt; [1] "CACNB2"  "ZBTB16"  "PRSS35"  "PDPN"    "SPARCL1" "DUSP1"</span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_de</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Subset data</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>significant <span class="op">=</span> <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">symbol</span> <span class="op">%in%</span> <span class="va">topgenes_symbols</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>, <span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="va">PValue</span>, label <span class="op">=</span> <span class="va">symbol</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">significant</span>, size <span class="op">=</span> <span class="va">significant</span>, alpha <span class="op">=</span> <span class="va">significant</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>

  <span class="co"># Custom scales</span>
  <span class="va">custom_theme</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10_reverse"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Scale for 'colour' is already present. Adding another scale for 'colour',</span>
<span class="co">#&gt; which will replace the existing scale.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-30-1.png" width="70%"></p>
</div>
<div id="automatic-bibliography" class="section level2">
<h2 class="hasAnchor">
<a href="#automatic-bibliography" class="anchor"></a>Automatic bibliography</h2>
<p>Tidybulk provides a handy function called <code>get_bibliography</code> that keeps track of the references for the methods used in your tidybulk workflow. The references are in BibTeX format and can be imported into your reference manager.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/get_bibliography-methods.html">get_bibliography</a></span><span class="op">(</span><span class="va">counts_de</span><span class="op">)</span>
<span class="co">#&gt;  @Article{tidybulk,</span>
<span class="co">#&gt;   title = {tidybulk: an R tidy framework for modular transcriptomic data analysis},</span>
<span class="co">#&gt;   author = {Stefano Mangiola and Ramyar Molania and Ruining Dong and Maria A. Doyle &amp; Anthony T. Papenfuss},</span>
<span class="co">#&gt;   journal = {Genome Biology},</span>
<span class="co">#&gt;   year = {2021},</span>
<span class="co">#&gt;   volume = {22},</span>
<span class="co">#&gt;   number = {42},</span>
<span class="co">#&gt;   url = {https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7},</span>
<span class="co">#&gt;   }</span>
<span class="co">#&gt; @article{wickham2019welcome,</span>
<span class="co">#&gt;   title={Welcome to the Tidyverse},</span>
<span class="co">#&gt;   author={Wickham, Hadley and Averick, Mara and Bryan, Jennifer and Chang, Winston and McGowan, Lucy D'Agostino and Francois, Romain and Grolemund, Garrett and Hayes, Alex and Henry, Lionel and Hester, Jim and others},</span>
<span class="co">#&gt;   journal={Journal of Open Source Software},</span>
<span class="co">#&gt;   volume={4},</span>
<span class="co">#&gt;   number={43},</span>
<span class="co">#&gt;   pages={1686},</span>
<span class="co">#&gt;   year={2019}</span>
<span class="co">#&gt;  }</span>
<span class="co">#&gt; @article{robinson2010edger,</span>
<span class="co">#&gt;   title={edgeR: a Bioconductor package for differential expression analysis of digital gene expression data},</span>
<span class="co">#&gt;   author={Robinson, Mark D and McCarthy, Davis J and Smyth, Gordon K},</span>
<span class="co">#&gt;   journal={Bioinformatics},</span>
<span class="co">#&gt;   volume={26},</span>
<span class="co">#&gt;   number={1},</span>
<span class="co">#&gt;   pages={139--140},</span>
<span class="co">#&gt;   year={2010},</span>
<span class="co">#&gt;   publisher={Oxford University Press}</span>
<span class="co">#&gt;  }</span>
<span class="co">#&gt; @article{robinson2010scaling,</span>
<span class="co">#&gt;   title={A scaling normalization method for differential expression analysis of RNA-seq data},</span>
<span class="co">#&gt;   author={Robinson, Mark D and Oshlack, Alicia},</span>
<span class="co">#&gt;   journal={Genome biology},</span>
<span class="co">#&gt;   volume={11},</span>
<span class="co">#&gt;   number={3},</span>
<span class="co">#&gt;   pages={1--9},</span>
<span class="co">#&gt;   year={2010},</span>
<span class="co">#&gt;   publisher={BioMed Central}</span>
<span class="co">#&gt;  }</span>
<span class="co">#&gt; @Manual{,</span>
<span class="co">#&gt;     title = {R: A Language and Environment for Statistical Computing},</span>
<span class="co">#&gt;     author = {{R Core Team}},</span>
<span class="co">#&gt;     organization = {R Foundation for Statistical Computing},</span>
<span class="co">#&gt;     address = {Vienna, Austria},</span>
<span class="co">#&gt;     year = {2020},</span>
<span class="co">#&gt;     url = {https://www.R-project.org/},</span>
<span class="co">#&gt;   }</span>
<span class="co">#&gt; @article{lund2012detecting,</span>
<span class="co">#&gt;   title={Detecting differential expression in RNA-sequence data using quasi-likelihood with shrunken dispersion estimates},</span>
<span class="co">#&gt;   author={Lund, Steven P and Nettleton, Dan and McCarthy, Davis J and Smyth, Gordon K},</span>
<span class="co">#&gt;   journal={Statistical applications in genetics and molecular biology},</span>
<span class="co">#&gt;   volume={11},</span>
<span class="co">#&gt;   number={5},</span>
<span class="co">#&gt;   year={2012},</span>
<span class="co">#&gt;   publisher={De Gruyter}</span>
<span class="co">#&gt;     }</span></code></pre></div>
</div>
<div id="cell-type-composition-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-composition-analysis" class="anchor"></a>Cell type composition analysis</h2>
<p>If we are sequencing tissue samples, we may want to know what cell types are present and if there are differences in expression between them. <code>tidybulk</code> has a <code>deconvolve_cellularity</code> function that can help us do this.</p>
<p>For this example we will use a subset of the breast cancer dataset from <a href="https://www.cancer.gov/tcga">The Cancer Genome Atlas (TCGA)</a>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA</span> <span class="op">&lt;-</span> <span class="fu">rladiestunis2021tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/BRCA.html">BRCA</a></span> 
<span class="va">BRCA</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 198,374 x 5</span>
<span class="co">#&gt; [90m# Transcripts=9017 | Samples=22 | Assays=count[39m</span>
<span class="co">#&gt;    feature  sample       count  time event_occurred</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt;        &lt;int&gt; &lt;dbl&gt;          &lt;int&gt;</span>
<span class="co">#&gt;  1 A1BG     TCGA-A2-A04P    13   102              1</span>
<span class="co">#&gt;  2 A1BG-AS1 TCGA-A2-A04P    61   102              1</span>
<span class="co">#&gt;  3 A2M      TCGA-A2-A04P 19277   102              1</span>
<span class="co">#&gt;  4 A2MP1    TCGA-A2-A04P     9   102              1</span>
<span class="co">#&gt;  5 AAAS     TCGA-A2-A04P  1917   102              1</span>
<span class="co">#&gt;  6 AACS     TCGA-A2-A04P  1998   102              1</span>
<span class="co">#&gt;  7 AADAT    TCGA-A2-A04P   212   102              1</span>
<span class="co">#&gt;  8 AAGAB    TCGA-A2-A04P  1194   102              1</span>
<span class="co">#&gt;  9 AAK1     TCGA-A2-A04P  1468   102              1</span>
<span class="co">#&gt; 10 AAMP     TCGA-A2-A04P  5798   102              1</span>
<span class="co">#&gt; # … with 40 more rows</span></code></pre></div>
<p>With tidybulk, we can easily infer the proportions of cell types within a tissue using one of several published methods (Cibersort <span class="citation">(Newman et al. 2015)</span>, EPIC <span class="citation">(Racle et al. 2017)</span> and llsr <span class="citation">(Abbas et al. 2009)</span>). Here we will use Cibersort which provides a default signature called LM22 to define the cell types. LM22 contains 547 genes that identify 22 human immune cell types.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type</span> <span class="op">&lt;-</span>
  <span class="va">BRCA</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/deconvolve_cellularity-methods.html">deconvolve_cellularity</a></span><span class="op">(</span>prefix<span class="op">=</span><span class="st">"cibersort: "</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>

<span class="va">BRCA_cell_type</span>
<span class="co">#&gt; # A SummarizedExperiment-tibble abstraction: 198,374 x 27</span>
<span class="co">#&gt; [90m# Transcripts=9017 | Samples=22 | Assays=count[39m</span>
<span class="co">#&gt;    feature sample count  time event_occurred cibersort..B.ce… cibersort..B.ce…</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt; &lt;dbl&gt;          &lt;int&gt;            &lt;dbl&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt;  1 A1BG    TCGA-…    13   102              1            0.201                0</span>
<span class="co">#&gt;  2 A1BG-A… TCGA-…    61   102              1            0.201                0</span>
<span class="co">#&gt;  3 A2M     TCGA-… 19277   102              1            0.201                0</span>
<span class="co">#&gt;  4 A2MP1   TCGA-…     9   102              1            0.201                0</span>
<span class="co">#&gt;  5 AAAS    TCGA-…  1917   102              1            0.201                0</span>
<span class="co">#&gt;  6 AACS    TCGA-…  1998   102              1            0.201                0</span>
<span class="co">#&gt;  7 AADAT   TCGA-…   212   102              1            0.201                0</span>
<span class="co">#&gt;  8 AAGAB   TCGA-…  1194   102              1            0.201                0</span>
<span class="co">#&gt;  9 AAK1    TCGA-…  1468   102              1            0.201                0</span>
<span class="co">#&gt; 10 AAMP    TCGA-…  5798   102              1            0.201                0</span>
<span class="co">#&gt; # … with 40 more rows, and 20 more variables: cibersort..Plasma.cells &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.CD8 &lt;dbl&gt;, cibersort..T.cells.CD4.naive &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.CD4.memory.resting &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.CD4.memory.activated &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.follicular.helper &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.regulatory..Tregs. &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..T.cells.gamma.delta &lt;dbl&gt;, cibersort..NK.cells.resting &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..NK.cells.activated &lt;dbl&gt;, cibersort..Monocytes &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Macrophages.M0 &lt;dbl&gt;, cibersort..Macrophages.M1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Macrophages.M2 &lt;dbl&gt;, cibersort..Dendritic.cells.resting &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Dendritic.cells.activated &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Mast.cells.resting &lt;dbl&gt;, cibersort..Mast.cells.activated &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cibersort..Eosinophils &lt;dbl&gt;, cibersort..Neutrophils &lt;dbl&gt;</span></code></pre></div>
<p>Cell type proportions are added to the tibble as new columns. The prefix makes it easy to reshape the data frame if needed, for visualisation or further analyses.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type_long</span> <span class="op">&lt;-</span>
  <span class="va">BRCA_cell_type</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Reshape</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/tidyr-methods.html">pivot_longer</a></span><span class="op">(</span>
    <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"cibersort"</span><span class="op">)</span>,
    names_prefix <span class="op">=</span> <span class="st">"cibersort: "</span>,
    names_to <span class="op">=</span> <span class="st">"cell_type"</span>,
    values_to <span class="op">=</span> <span class="st">"proportion"</span>
  <span class="op">)</span>

<span class="va">BRCA_cell_type_long</span>
<span class="co">#&gt; # A tibble: 484 x 5</span>
<span class="co">#&gt;    sample       time event_occurred cell_type                         proportion</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;dbl&gt;          &lt;int&gt; &lt;chr&gt;                                  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA-A2-A0…   102              1 cibersort..B.cells.naive              0.201 </span>
<span class="co">#&gt;  2 TCGA-A2-A0…   102              1 cibersort..B.cells.memory             0     </span>
<span class="co">#&gt;  3 TCGA-A2-A0…   102              1 cibersort..Plasma.cells               0     </span>
<span class="co">#&gt;  4 TCGA-A2-A0…   102              1 cibersort..T.cells.CD8                0.0297</span>
<span class="co">#&gt;  5 TCGA-A2-A0…   102              1 cibersort..T.cells.CD4.naive          0     </span>
<span class="co">#&gt;  6 TCGA-A2-A0…   102              1 cibersort..T.cells.CD4.memory.re…     0.0841</span>
<span class="co">#&gt;  7 TCGA-A2-A0…   102              1 cibersort..T.cells.CD4.memory.ac…     0     </span>
<span class="co">#&gt;  8 TCGA-A2-A0…   102              1 cibersort..T.cells.follicular.he…     0.0556</span>
<span class="co">#&gt;  9 TCGA-A2-A0…   102              1 cibersort..T.cells.regulatory..T…     0.0484</span>
<span class="co">#&gt; 10 TCGA-A2-A0…   102              1 cibersort..T.cells.gamma.delta        0     </span>
<span class="co">#&gt; # … with 474 more rows</span></code></pre></div>
<p>We can plot the proportions of immune cell types for each patient.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">BRCA_cell_type_long</span> <span class="op">%&gt;%</span>

  <span class="co"># Plot proportions</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySummarizedExperiment/man/ggplot2-methods.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>, y <span class="op">=</span> <span class="va">proportion</span>, fill <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
</div>
<div id="key-points" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points" class="anchor"></a>Key Points</h2>
<ul>
<li>Bulk RNA sequencing data can be represented and analysed in a ‘tidy’ way using tidySummarizedExperiment, tidybulk and the tidyverse.</li>
<li>tidySummarizedExperiment enables us to visualise and manipulate a Bioconductor SummarizedExperiment object as if it were in tidy data format.</li>
<li>Some of the key steps in an RNA sequencing analysis are filtering lowly abundant transcripts, adjusting for differences in sequencing depth and composition, testing for differential expression, and visualising the data, which can all be performed in a tidy way with tidybulk.</li>
<li>
<code>tidybulk</code> allows streamlined multi-method analyses</li>
<li>
<code>tidybulk</code> allow easy analyses of cell type composition</li>
</ul>
</div>
</div>
<div id="part-2-single-cell-rna-seq-with-tidysinglecellexperiment" class="section level1">
<h1 class="hasAnchor">
<a href="#part-2-single-cell-rna-seq-with-tidysinglecellexperiment" class="anchor"></a>Part 2 Single-cell RNA-seq with tidySingleCellExperiment</h1>
<p>A typical scRNA-seq workflow is shown in <a href="https://stemangiola.github.io/rladiestunis2021_tidytranscriptomics/articles/background.html#differences-between-bulk-and-single-cell-rna-sequencing-1">Background information</a>. We don’t have time in this workshop to go into depth on each step but you can read more about scRNA-seq workflows in the online book <a href="http://bioconductor.org/books/release/OSCA/index.html">Orchestrating Single-Cell Analysis with Bioconductor</a>.</p>
<p>In Part 1 we showed how we can study the cell-type composition of a biological sample using bulk RNA sequencing. Single cell sequencing enables a more direct estimation of cell-type composition and gives greater resolution. For bulk RNA sequencing we need to infer the cell types using the abundance of transcripts in the whole sample, with single-cell RNA sequencing we can directly measure the transcripts in each cell and then classify the cells into cell types.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org">igraph</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR">SingleR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex">celldex</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm">ggbeeswarm</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidySingleCellExperiment</span><span class="op">)</span></code></pre></div>
<div id="introduction-to-tidysinglecellexperiment" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-tidysinglecellexperiment" class="anchor"></a>Introduction to tidySingleCellExperiment</h2>
<p>The single-cell RNA sequencing data used here is 3000 cells in total, subsetted from 20 samples from 10 peripheral blood mononuclear cell (pbmc) datasets. The datasets are from GSE115189/SRR7244582 <span class="citation">(Freytag et al. 2018)</span>, SRR11038995 [<span class="citation">Cai et al. (2020)</span>, SCP345 (singlecell.broadinstitute.org), SCP424 <span class="citation">(Ding et al. 2020)</span>, SCP591 <span class="citation">(Karagiannis et al. 2020)</span> and 10x-derived 6K and 8K datasets (support.10xgenomics.com/). The data is in <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment</a> format. SingleCellExperiment is a very popular container of single cell RNA sequencing data.</p>
<p>Similar to what we saw with tidySummarizedExperiment, <code>tidySingleCellExperiment</code> package enables any <em>SingleCellExperiment</em> object to be displayed and manipulated according to tidy data principles, without affecting any <em>SingleCellExperiment</em> behaviour.</p>
<p>If we load the <em>tidySingleCellExperiment</em> package and then view the pbmc data it displays as a tibble.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load pbmc single cell RNA sequencing data</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">rladiestunis2021tidytranscriptomics</span><span class="fu">::</span><span class="va"><a href="../reference/pbmc.html">pbmc</a></span>

<span class="co"># take a look</span>
<span class="va">pbmc</span>
<span class="co">#&gt; # A SingleCellExperiment-tibble abstraction: 3,000 x 8</span>
<span class="co">#&gt; [90m# Transcripts=51958 | Assays=counts, logcounts[39m</span>
<span class="co">#&gt;    cell    file      orig.ident nCount_RNA nFeature_RNA  S.Score G2M.Score ident</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;           &lt;dbl&gt;        &lt;int&gt;    &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;</span>
<span class="co">#&gt;  1 CCAGTC… ../data/… SeuratPro…       3421          979 -5.42e-2  -0.0107  G1   </span>
<span class="co">#&gt;  2 ATGAGC… ../data/… SeuratPro…       2752          898 -5.01e-2  -0.00416 G1   </span>
<span class="co">#&gt;  3 TATGAA… ../data/… SeuratPro…       2114          937 -2.95e-5  -0.0229  G1   </span>
<span class="co">#&gt;  4 CATATA… ../data/… SeuratPro…       3122         1086 -6.65e-2  -0.0488  G1   </span>
<span class="co">#&gt;  5 GAGGCA… ../data/… SeuratPro…       2341          957 -3.74e-3   0.0241  G2M  </span>
<span class="co">#&gt;  6 AGCTGC… ../data/… SeuratPro…       5472         1758 -5.88e-2   0.00241 G2M  </span>
<span class="co">#&gt;  7 TGATTA… ../data/… SeuratPro…       1258          542 -2.51e-2  -0.0269  G1   </span>
<span class="co">#&gt;  8 ACGAAG… ../data/… SeuratPro…       7683         1926 -1.33e-1  -0.116   G1   </span>
<span class="co">#&gt;  9 CGGCAT… ../data/… SeuratPro…       3500         1092 -6.87e-2  -0.0622  G1   </span>
<span class="co">#&gt; 10 ATAGCG… ../data/… SeuratPro…       3092          974 -1.24e-2  -0.0271  G1   </span>
<span class="co">#&gt; # … with 2,990 more rows</span></code></pre></div>
<p>It can be interacted with using <a href="https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html">SingleCellExperiment commands</a> such as <code>assayNames</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assayNames</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>
<span class="co">#&gt; [1] "counts"    "logcounts"</span></code></pre></div>
<p>We can also interact with our object as we do with any tidyverse tibble.</p>
</div>
<div id="setting-up-the-data-1" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-data-1" class="anchor"></a>Setting up the data</h2>
<p>In this case we want to polish an annotation column. We will extract the sample, dataset and group information from the file name column into separate columns.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First take a look at the file column</span>
<span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span>
<span class="co">#&gt; tidySingleCellExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 3,000 x 1</span>
<span class="co">#&gt;    file                                                </span>
<span class="co">#&gt;    &lt;chr&gt;                                               </span>
<span class="co">#&gt;  1 ../data/GSE115189_1/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt;  2 ../data/GSE115189_2/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt;  3 ../data/GSE115189_2/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt;  4 ../data/GSE115189_1/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt;  5 ../data/GSE115189_2/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt;  6 ../data/GSE115189_1/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt;  7 ../data/GSE115189_1/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt;  8 ../data/GSE115189_1/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt;  9 ../data/GSE115189_1/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt; 10 ../data/GSE115189_1/outs/filtered_feature_bc_matrix/</span>
<span class="co">#&gt; # … with 2,990 more rows</span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create columns for sample, dataset and groups</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span>
  <span class="va">pbmc</span> <span class="op">%&gt;%</span>

  <span class="co"># Extract sample and group</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">extract</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/([a-zA-Z0-9_]+)/outs.+"</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Extract data source</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">extract</a></span><span class="op">(</span><span class="va">file</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dataset"</span>, <span class="st">"groups"</span><span class="op">)</span>, <span class="st">"../data/([a-zA-Z0-9_]+)_([0-9])/outs.+"</span><span class="op">)</span>

<span class="co"># Take a look</span>
<span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">dataset</span>, <span class="va">groups</span><span class="op">)</span>
<span class="co">#&gt; tidySingleCellExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 3,000 x 3</span>
<span class="co">#&gt;    sample      dataset   groups</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt; </span>
<span class="co">#&gt;  1 GSE115189_1 GSE115189 1     </span>
<span class="co">#&gt;  2 GSE115189_2 GSE115189 2     </span>
<span class="co">#&gt;  3 GSE115189_2 GSE115189 2     </span>
<span class="co">#&gt;  4 GSE115189_1 GSE115189 1     </span>
<span class="co">#&gt;  5 GSE115189_2 GSE115189 2     </span>
<span class="co">#&gt;  6 GSE115189_1 GSE115189 1     </span>
<span class="co">#&gt;  7 GSE115189_1 GSE115189 1     </span>
<span class="co">#&gt;  8 GSE115189_1 GSE115189 1     </span>
<span class="co">#&gt;  9 GSE115189_1 GSE115189 1     </span>
<span class="co">#&gt; 10 GSE115189_1 GSE115189 1     </span>
<span class="co">#&gt; # … with 2,990 more rows</span></code></pre></div>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h2>
<p>A key quality control step performed in single cell analysis is assessment of the proportion of mitochondrial transcripts. A high mitochondrial count can indicate the cell is dead or dying and is thus poor quality and should be filtered out.</p>
<div id="nesting" class="section level3">
<h3 class="hasAnchor">
<a href="#nesting" class="anchor"></a>Nesting</h3>
<p>A powerful tool tidySingleCellExperiment enables us to use with our single-cell data is tidyverse’s <a href="https://rstudio-education.github.io/tidyverse-cookbook/transform-tables.html#nest-a-data-frame">nest</a>. Nesting allows us to easily perform independent analyses on subsets of the data. For example, our data contains 10 single-cell datasets from different sources, we can nest by dataset and analyse the mitochondrial content for each.</p>
<p>To demonstrate the power of nesting we’ll first show the mitochondrial analysis for one of the 10 datasets.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">one_dataset</span> <span class="op">&lt;-</span> <span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">filter</a></span><span class="op">(</span><span class="va">dataset</span> <span class="op">==</span><span class="st">"GSE115189"</span><span class="op">)</span></code></pre></div>
<p>We get the chromosomal location for each gene in the dataset so we can identify the mitochondrial genes. We’ll get a warning that some of the ids don’t find a match but it should be just a small proportion.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">location</span> <span class="op">&lt;-</span> <span class="fu">mapIds</span><span class="op">(</span>
  <span class="va">EnsDb.Hsapiens.v86</span>,
  keys <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">one_dataset</span><span class="op">)</span>,
  column <span class="op">=</span> <span class="st">"SEQNAME"</span>,
  keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>
<span class="op">)</span>
<span class="co">#&gt; Warning: Unable to map 5313 of 51958 requested IDs.</span></code></pre></div>
<p>Next we calculate the mitchondrial content for each cell in the dataset.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mito_info_one_dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html">perCellQCMetrics</a></span><span class="op">(</span><span class="va">one_dataset</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Mito <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">location</span> <span class="op">==</span> <span class="st">"MT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">mito_info_one_dataset</span>
<span class="co">#&gt; DataFrame with 300 rows and 6 columns</span>
<span class="co">#&gt;                        sum  detected subsets_Mito_sum subsets_Mito_detected</span>
<span class="co">#&gt;                  &lt;numeric&gt; &lt;integer&gt;        &lt;numeric&gt;             &lt;integer&gt;</span>
<span class="co">#&gt; CCAGTCACACTGGT-1      3421       979              132                    11</span>
<span class="co">#&gt; ATGAGCACATCTTC-1      2752       898               33                     9</span>
<span class="co">#&gt; TATGAATGGAGGAC-1      2114       937               21                     7</span>
<span class="co">#&gt; CATATAGACTAAGC-1      3122      1086               54                    10</span>
<span class="co">#&gt; GAGGCAGACTTGCC-1      2341       957               30                    10</span>
<span class="co">#&gt; ...                    ...       ...              ...                   ...</span>
<span class="co">#&gt; GATATCCTAGAAGT-1      2432       925               34                    10</span>
<span class="co">#&gt; GAACCAACTTCCGC-1      1682       819               34                     9</span>
<span class="co">#&gt; CGACGTCTGAGGCA-1      1136       448               16                     8</span>
<span class="co">#&gt; TCAGCAGACTCCAC-1      2807       931               30                     9</span>
<span class="co">#&gt; AACACGTGTACGAC-1      3743      1250              105                    12</span>
<span class="co">#&gt;                  subsets_Mito_percent     total</span>
<span class="co">#&gt;                             &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">#&gt; CCAGTCACACTGGT-1             3.858521      3421</span>
<span class="co">#&gt; ATGAGCACATCTTC-1             1.199128      2752</span>
<span class="co">#&gt; TATGAATGGAGGAC-1             0.993377      2114</span>
<span class="co">#&gt; CATATAGACTAAGC-1             1.729660      3122</span>
<span class="co">#&gt; GAGGCAGACTTGCC-1             1.281504      2341</span>
<span class="co">#&gt; ...                               ...       ...</span>
<span class="co">#&gt; GATATCCTAGAAGT-1              1.39803      2432</span>
<span class="co">#&gt; GAACCAACTTCCGC-1              2.02140      1682</span>
<span class="co">#&gt; CGACGTCTGAGGCA-1              1.40845      1136</span>
<span class="co">#&gt; TCAGCAGACTCCAC-1              1.06876      2807</span>
<span class="co">#&gt; AACACGTGTACGAC-1              2.80524      3743</span></code></pre></div>
<p>We then label the cells with high mitochondrial content as outliers.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mito_info_one_dataset</span> <span class="op">&lt;-</span> <span class="va">mito_info_one_dataset</span> <span class="op">%&gt;%</span>     
      <span class="co"># Converting to tibble</span>
      <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tibble-methods.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span> <span class="op">%&gt;%</span>

      <span class="co"># Label cells with high mitochondrial content</span>
      <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>high_mitochondrion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html">isOutlier</a></span><span class="op">(</span><span class="va">subsets_Mito_percent</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span><span class="op">)</span>

<span class="va">mito_info_one_dataset</span>
<span class="co">#&gt; # A tibble: 300 x 8</span>
<span class="co">#&gt;    cell    sum detected subsets_Mito_sum subsets_Mito_de… subsets_Mito_pe… total</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;dbl&gt;    &lt;int&gt;            &lt;dbl&gt;            &lt;int&gt;            &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 CCAG…  3421      979              132               11            3.86   3421</span>
<span class="co">#&gt;  2 ATGA…  2752      898               33                9            1.20   2752</span>
<span class="co">#&gt;  3 TATG…  2114      937               21                7            0.993  2114</span>
<span class="co">#&gt;  4 CATA…  3122     1086               54               10            1.73   3122</span>
<span class="co">#&gt;  5 GAGG…  2341      957               30               10            1.28   2341</span>
<span class="co">#&gt;  6 AGCT…  5472     1758              169               13            3.09   5472</span>
<span class="co">#&gt;  7 TGAT…  1258      542               41                9            3.26   1258</span>
<span class="co">#&gt;  8 ACGA…  7683     1926              276               14            3.59   7683</span>
<span class="co">#&gt;  9 CGGC…  3500     1092               75               10            2.14   3500</span>
<span class="co">#&gt; 10 ATAG…  3092      974               37               10            1.20   3092</span>
<span class="co">#&gt; # … with 290 more rows, and 1 more variable: high_mitochondrion &lt;lgl&gt;</span></code></pre></div>
<p>Finally, we join the mitochondrial information back to the original data so we will be able to filter out the cells with high mitochondrial content.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mito_info_one_dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">left_join</a></span><span class="op">(</span><span class="va">one_dataset</span>, <span class="va">mito_info_one_dataset</span>, by <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span>

<span class="va">mito_info_one_dataset</span>
<span class="co">#&gt; # A SingleCellExperiment-tibble abstraction: 300 x 17</span>
<span class="co">#&gt; [90m# Transcripts=51958 | Assays=counts, logcounts[39m</span>
<span class="co">#&gt;    cell  dataset groups sample orig.ident nCount_RNA nFeature_RNA  S.Score</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;           &lt;dbl&gt;        &lt;int&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1 CCAG… GSE115… 1      GSE11… SeuratPro…       3421          979 -5.42e-2</span>
<span class="co">#&gt;  2 ATGA… GSE115… 2      GSE11… SeuratPro…       2752          898 -5.01e-2</span>
<span class="co">#&gt;  3 TATG… GSE115… 2      GSE11… SeuratPro…       2114          937 -2.95e-5</span>
<span class="co">#&gt;  4 CATA… GSE115… 1      GSE11… SeuratPro…       3122         1086 -6.65e-2</span>
<span class="co">#&gt;  5 GAGG… GSE115… 2      GSE11… SeuratPro…       2341          957 -3.74e-3</span>
<span class="co">#&gt;  6 AGCT… GSE115… 1      GSE11… SeuratPro…       5472         1758 -5.88e-2</span>
<span class="co">#&gt;  7 TGAT… GSE115… 1      GSE11… SeuratPro…       1258          542 -2.51e-2</span>
<span class="co">#&gt;  8 ACGA… GSE115… 1      GSE11… SeuratPro…       7683         1926 -1.33e-1</span>
<span class="co">#&gt;  9 CGGC… GSE115… 1      GSE11… SeuratPro…       3500         1092 -6.87e-2</span>
<span class="co">#&gt; 10 ATAG… GSE115… 1      GSE11… SeuratPro…       3092          974 -1.24e-2</span>
<span class="co">#&gt; # … with 290 more rows, and 9 more variables: G2M.Score &lt;dbl&gt;, ident &lt;fct&gt;,</span>
<span class="co">#&gt; #   sum &lt;dbl&gt;, detected &lt;int&gt;, subsets_Mito_sum &lt;dbl&gt;,</span>
<span class="co">#&gt; #   subsets_Mito_detected &lt;int&gt;, subsets_Mito_percent &lt;dbl&gt;, total &lt;dbl&gt;,</span>
<span class="co">#&gt; #   high_mitochondrion &lt;lgl&gt;</span></code></pre></div>
<p>The steps above perform the mitochondrial QC analysis for one dataset. To analyse all 10 datasets we could repeat the steps for each dataset or create a function and apply it to each dataset. However, a more efficient way is to use nesting. We can nest our data by dataset.</p>
<p><img src="../inst/vignettes/nesting.png" width="600px"></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_nested</span> <span class="op">&lt;-</span> <span class="va">pbmc</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">dataset</span><span class="op">)</span>

<span class="va">pbmc_nested</span>
<span class="co">#&gt; # A tibble: 10 x 2</span>
<span class="co">#&gt;    dataset      data            </span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;list&gt;          </span>
<span class="co">#&gt;  1 GSE115189    &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  2 10x_6K       &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  3 10x_8K       &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  4 SRR11038995  &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  5 SRR7244582   &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  6 SCP345_580   &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  7 SCP345_860   &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  8 SCP424_pbmc1 &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  9 SCP424_pbmc2 &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt; 10 SCP591       &lt;SnglCllE[,300]&gt;</span></code></pre></div>
<p>We will create a table (tibble) with the mitochondrial QC information for each dataset. Here we use <code>map</code> which is often used in combination with <code>nest</code> to perform commands on nested data.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">location</span> <span class="op">&lt;-</span> <span class="fu">mapIds</span><span class="op">(</span>
  <span class="va">EnsDb.Hsapiens.v86</span>,
  keys <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,
  column <span class="op">=</span> <span class="st">"SEQNAME"</span>,
  keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>
<span class="op">)</span>
<span class="co">#&gt; Warning: Unable to map 5313 of 51958 requested IDs.</span>

<span class="va">mito_info_all_datasets</span> <span class="op">&lt;-</span> <span class="va">pbmc_nested</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>mitochondrion_info <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span>
    <span class="va">data</span>,
    <span class="op">~</span> <span class="co"># Calculate mitochondrial statistics</span>
      <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html">perCellQCMetrics</a></span><span class="op">(</span><span class="va">.x</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Mito <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">location</span> <span class="op">==</span> <span class="st">"MT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
      
      <span class="co"># Convert to tibble</span>
      <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tibble-methods.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span> <span class="op">%&gt;%</span>

      <span class="co"># Label cells with high mitochondrial content</span>
      <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>high_mitochondrion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html">isOutlier</a></span><span class="op">(</span><span class="va">subsets_Mito_percent</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span><span class="op">)</span>
  
<span class="va">mito_info_all_datasets</span>
<span class="co">#&gt; # A tibble: 10 x 3</span>
<span class="co">#&gt;    dataset      data             mitochondrion_info    </span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;list&gt;           &lt;list&gt;                </span>
<span class="co">#&gt;  1 GSE115189    &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt;  2 10x_6K       &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt;  3 10x_8K       &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt;  4 SRR11038995  &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt;  5 SRR7244582   &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt;  6 SCP345_580   &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt;  7 SCP345_860   &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt;  8 SCP424_pbmc1 &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt;  9 SCP424_pbmc2 &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span>
<span class="co">#&gt; 10 SCP591       &lt;SnglCllE[,300]&gt; &lt;tibble[,8] [300 × 8]&gt;</span></code></pre></div>
<p>As before, we join the mitochondrial information to the original data.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="co"># Join the mitochondrial information to the original SingleCellExperiment data</span>
 <span class="va">mito_info_all_datasets</span> <span class="op">&lt;-</span> <span class="va">mito_info_all_datasets</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html">map2</a></span><span class="op">(</span>
    <span class="va">data</span>, <span class="va">mitochondrion_info</span>,
    <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">left_join</a></span><span class="op">(</span><span class="va">.x</span>, <span class="va">.y</span>, by <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span>
  <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Remove the separate mitochondrial information table</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">mitochondrion_info</span><span class="op">)</span>

<span class="va">mito_info_all_datasets</span>
<span class="co">#&gt; # A tibble: 10 x 2</span>
<span class="co">#&gt;    dataset      data            </span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;list&gt;          </span>
<span class="co">#&gt;  1 GSE115189    &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  2 10x_6K       &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  3 10x_8K       &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  4 SRR11038995  &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  5 SRR7244582   &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  6 SCP345_580   &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  7 SCP345_860   &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  8 SCP424_pbmc1 &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt;  9 SCP424_pbmc2 &lt;SnglCllE[,300]&gt;</span>
<span class="co">#&gt; 10 SCP591       &lt;SnglCllE[,300]&gt;</span></code></pre></div>
<p>Remove the nesting to get back one table containing all datasets.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="va">mito_info_all_datasets</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>

<span class="va">pbmc</span>
<span class="co">#&gt; # A SingleCellExperiment-tibble abstraction: 3,000 x 17</span>
<span class="co">#&gt; [90m# Transcripts=51958 | Assays=counts, logcounts[39m</span>
<span class="co">#&gt;    cell  groups sample orig.ident nCount_RNA nFeature_RNA  S.Score G2M.Score</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;           &lt;dbl&gt;        &lt;int&gt;    &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 CCAG… 1      GSE11… SeuratPro…       3421          979 -5.42e-2  -0.0107 </span>
<span class="co">#&gt;  2 ATGA… 2      GSE11… SeuratPro…       2752          898 -5.01e-2  -0.00416</span>
<span class="co">#&gt;  3 TATG… 2      GSE11… SeuratPro…       2114          937 -2.95e-5  -0.0229 </span>
<span class="co">#&gt;  4 CATA… 1      GSE11… SeuratPro…       3122         1086 -6.65e-2  -0.0488 </span>
<span class="co">#&gt;  5 GAGG… 2      GSE11… SeuratPro…       2341          957 -3.74e-3   0.0241 </span>
<span class="co">#&gt;  6 AGCT… 1      GSE11… SeuratPro…       5472         1758 -5.88e-2   0.00241</span>
<span class="co">#&gt;  7 TGAT… 1      GSE11… SeuratPro…       1258          542 -2.51e-2  -0.0269 </span>
<span class="co">#&gt;  8 ACGA… 1      GSE11… SeuratPro…       7683         1926 -1.33e-1  -0.116  </span>
<span class="co">#&gt;  9 CGGC… 1      GSE11… SeuratPro…       3500         1092 -6.87e-2  -0.0622 </span>
<span class="co">#&gt; 10 ATAG… 1      GSE11… SeuratPro…       3092          974 -1.24e-2  -0.0271 </span>
<span class="co">#&gt; # … with 2,990 more rows, and 9 more variables: ident &lt;fct&gt;, sum &lt;dbl&gt;,</span>
<span class="co">#&gt; #   detected &lt;int&gt;, subsets_Mito_sum &lt;dbl&gt;, subsets_Mito_detected &lt;int&gt;,</span>
<span class="co">#&gt; #   subsets_Mito_percent &lt;dbl&gt;, total &lt;dbl&gt;, high_mitochondrion &lt;lgl&gt;,</span>
<span class="co">#&gt; #   dataset &lt;chr&gt;</span></code></pre></div>
<p>Nesting has allowed us to perform the QC on each dataset in an efficient way.</p>
<p>We can use tidyverse to reshape the data and create <a href="http://www.cbs.dtu.dk/~eklund/beeswarm/">beeswarm plots</a> to visualise the mitochondrial content.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">%&gt;%</span>

  <span class="co"># Reshaping</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/pivot-methods.html">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">detected</span>, <span class="va">sum</span>, <span class="va">subsets_Mito_percent</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="va">dataset</span>, y <span class="op">=</span> <span class="va">value</span>,
    color <span class="op">=</span> <span class="va">high_mitochondrion</span>,
    alpha <span class="op">=</span> <span class="va">high_mitochondrion</span>,
    size <span class="op">=</span> <span class="va">high_mitochondrion</span>
  <span class="op">)</span><span class="op">)</span> <span class="op">+</span>

  <span class="co"># Plotting</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_quasirandom.html">geom_quasirandom</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">name</span>, scale <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span>

  <span class="co"># Customisation</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">50</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span></code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<p>In the faceted plot, detected is number of genes in each of the 10 datasets, sum is total counts.</p>
<p>We then proceed to filter out cells with high mitochondrial content.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">high_mitochondrion</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="scaling-and-integrating" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-and-integrating" class="anchor"></a>Scaling and Integrating</h2>
<p>As we are working with multiple datasets, we need to integrate them and adjust for technical variability between them. Here we’ll nest by dataset (batch), normalise within each batch with <code>multiBatchNorm</code> and correct for batch effects with <code>fastMNN</code>.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Scaling within each dataset</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span>
  <span class="va">pbmc</span> <span class="op">%&gt;%</span>

  <span class="co"># Define batch</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>batch <span class="op">=</span> <span class="va">dataset</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidyr-methods.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">batch</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Normalisation   </span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">mutate</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html">multiBatchNorm</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>

  <span class="co"># Integration</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">pull</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html">fastMNN</a></span><span class="op">(</span><span class="op">)</span> </code></pre></div>
</div>
<div id="identify-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h2>
<p>We proceed with identifying cell clusters.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Assign clusters to the 'colLabels'</span>
<span class="co"># of the SingleCellExperiment object</span>
<span class="fu">colLabels</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="co"># from SingleCellExperiment</span>
  <span class="va">pbmc</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span><span class="op">(</span>use.dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># from scran - shared nearest neighbour</span>
  <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html">cluster_walktrap</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%$%</span> <span class="co"># from igraph</span>
  <span class="va">membership</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Reorder columns</span>
<span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">label</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A SingleCellExperiment-tibble abstraction: 2,899 x 8</span>
<span class="co">#&gt; [90m# Transcripts=51958 | Assays=reconstructed[39m</span>
<span class="co">#&gt;    cell       label batch corrected1 corrected2 corrected3 corrected4 corrected5</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;fct&gt; &lt;int&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 CCAGTCACA… 4         1   -0.0644      0.334     -0.0725     0.0560    0.0739 </span>
<span class="co">#&gt;  2 ATGAGCACA… 2         1   -0.178      -0.0167    -0.122      0.122    -0.0233 </span>
<span class="co">#&gt;  3 TATGAATGG… 6         1   -0.0604      0.0357     0.127      0.293    -0.0128 </span>
<span class="co">#&gt;  4 CATATAGAC… 11        1   -0.0760      0.282     -0.121      0.0545    0.00285</span>
<span class="co">#&gt;  5 GAGGCAGAC… 1         1   -0.0758      0.0424     0.0977     0.259    -0.0248 </span>
<span class="co">#&gt;  6 AGCTGCCTT… 5         1   -0.117       0.0242    -0.0729     0.159    -0.0170 </span>
<span class="co">#&gt;  7 TGATTAGAT… 3         1   -0.146       0.0752    -0.143      0.0488   -0.0915 </span>
<span class="co">#&gt;  8 ACGAAGCTC… 3         1   -0.119       0.108     -0.127      0.0505   -0.0624 </span>
<span class="co">#&gt;  9 CGGCATCTT… 4         1   -0.00948     0.407     -0.0466     0.0803    0.0522 </span>
<span class="co">#&gt; 10 ATAGCGTGC… 4         1   -0.0561      0.367     -0.105      0.0589    0.0610 </span>
<span class="co">#&gt; # … with 2,889 more rows</span></code></pre></div>
<p>Thanks to tidySingleCellExperiment we can interrogate the object with tidyverse commands and use <code>count</code> to count the number of cells in each cluster.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/count.html">count</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span>
<span class="co">#&gt; tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 11 x 2</span>
<span class="co">#&gt;    label     n</span>
<span class="co">#&gt;    &lt;fct&gt; &lt;int&gt;</span>
<span class="co">#&gt;  1 1       460</span>
<span class="co">#&gt;  2 2       795</span>
<span class="co">#&gt;  3 3       367</span>
<span class="co">#&gt;  4 4       339</span>
<span class="co">#&gt;  5 5       486</span>
<span class="co">#&gt;  6 6       276</span>
<span class="co">#&gt;  7 7        21</span>
<span class="co">#&gt;  8 8        74</span>
<span class="co">#&gt;  9 9        41</span>
<span class="co">#&gt; 10 10       14</span>
<span class="co">#&gt; 11 11       26</span></code></pre></div>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h2>
<p>Besides PCA which is a linear dimensionality reduction, we can apply neighbour aware methods such as UMAP, to better define locally similar cells. We can calculate the first 3 UMAP dimensions using the scater framework.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Calculate UMAP with scater</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span>
  <span class="va">pbmc</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span>ncomponents <span class="op">=</span> <span class="fl">3</span>, dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span> <span class="co"># from scater</span></code></pre></div>
<p>And we can plot the clusters as a 3D plot using plotly. This time we are colouring by estimated cluster labels to visually check the cluster labels.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/plot_ly.html">plot_ly</a></span><span class="op">(</span>
    x <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP1`</span>,
    y <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP2`</span>,
    z <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP3`</span>,
    colors <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,
    color <span class="op">=</span> <span class="op">~</span><span class="va">label</span>,
    size<span class="op">=</span><span class="fl">0.5</span> 
  <span class="op">)</span></code></pre></div>
<p><img src="../inst/vignettes/plotly_2.png" width="261"></p>
</div>
<div id="cell-type-classification" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-classification" class="anchor"></a>Cell type classification</h2>
<div id="manual-cell-type-classification" class="section level3">
<h3 class="hasAnchor">
<a href="#manual-cell-type-classification" class="anchor"></a>Manual cell type classification</h3>
<p>We can identify cluster markers (genes). As example, we are selecting the top 10 for each cluster. We can then plot a heatmap of those gene markers across cells.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Identify top 10 markers per cluster</span>
<span class="va">marker_genes</span> <span class="op">&lt;-</span>
    <span class="va">pbmc</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers</a></span><span class="op">(</span>groups<span class="op">=</span><span class="va">pbmc</span><span class="op">$</span><span class="va">label</span>, assay.type <span class="op">=</span> <span class="st">"reconstructed"</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="co"># scran</span>
    <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">.x</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Plot heatmap</span>
<span class="va">pbmc</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span>                                  <span class="co"># from scater</span>
    features<span class="op">=</span><span class="va">marker_genes</span>, 
    columns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">label</span><span class="op">)</span>, 
    colour_columns_by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"label"</span><span class="op">)</span>, 
    exprs_values  <span class="op">=</span> <span class="st">"reconstructed"</span>
  <span class="op">)</span> </code></pre></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-54-1.png" width="700"></p>
</div>
<div id="automatic-cell-type-classification" class="section level3">
<h3 class="hasAnchor">
<a href="#automatic-cell-type-classification" class="anchor"></a>Automatic cell type classification</h3>
<p>We can infer cell type identities (e.g. T cell) using <em>SingleR</em> <span class="citation">(Aran et al. 2019)</span> and manipulate the output using tidyverse. SingleR <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html">accepts any log-normalised transcript abundance matrix</a></p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Reference cell types</span>
<span class="va">blueprint</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/datasets.html">BlueprintEncodeData</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; using temporary cache /tmp/Rtmp78FZMV/BiocFileCache</span>
<span class="co">#&gt; snapshotDate(): 2020-10-27</span>
<span class="co">#&gt; see ?celldex and browseVignettes('celldex') for documentation</span>
<span class="co">#&gt; downloading 1 resources</span>
<span class="co">#&gt; retrieving 1 resource</span>
<span class="co">#&gt; loading from cache</span>
<span class="co">#&gt; see ?celldex and browseVignettes('celldex') for documentation</span>
<span class="co">#&gt; downloading 1 resources</span>
<span class="co">#&gt; retrieving 1 resource</span>
<span class="co">#&gt; loading from cache</span>

<span class="va">cell_type</span> <span class="op">&lt;-</span>
    
  <span class="co"># extracting counts from SingleCellExperiment object</span>
  <span class="fu">assays</span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">reconstructed</span> <span class="op">%&gt;%</span>

    <span class="co"># SingleR</span>
  <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html">SingleR</a></span><span class="op">(</span>
    ref <span class="op">=</span> <span class="va">blueprint</span>,
    labels <span class="op">=</span> <span class="va">blueprint</span><span class="op">$</span><span class="va">label.main</span>,
    clusters <span class="op">=</span> <span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">pull</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
    
  <span class="co"># Formatting results</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tibble-methods.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">first.labels</span><span class="op">)</span>

<span class="va">cell_type</span>
<span class="co">#&gt; # A tibble: 11 x 2</span>
<span class="co">#&gt;    label first.labels   </span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;          </span>
<span class="co">#&gt;  1 1     NK cells       </span>
<span class="co">#&gt;  2 2     Skeletal muscle</span>
<span class="co">#&gt;  3 3     B-cells        </span>
<span class="co">#&gt;  4 4     Monocytes      </span>
<span class="co">#&gt;  5 5     Fibroblasts    </span>
<span class="co">#&gt;  6 6     NK cells       </span>
<span class="co">#&gt;  7 7     Smooth muscle  </span>
<span class="co">#&gt;  8 8     Neutrophils    </span>
<span class="co">#&gt;  9 9     Macrophages    </span>
<span class="co">#&gt; 10 10    B-cells        </span>
<span class="co">#&gt; 11 11    Monocytes</span></code></pre></div>
<p>We join the cell type information to our pbmc data.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Join cell type info</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span>
  <span class="va">pbmc</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">left_join</a></span><span class="op">(</span><span class="va">cell_type</span>, by <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span>

<span class="co"># Reorder columns</span>
<span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/dplyr-methods.html">select</a></span><span class="op">(</span><span class="va">cell</span>, <span class="va">first.labels</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A SingleCellExperiment-tibble abstraction: 2,899 x 12</span>
<span class="co">#&gt; [90m# Transcripts=51958 | Assays=reconstructed[39m</span>
<span class="co">#&gt;    cell  first.labels batch label corrected1 corrected2 corrected3 corrected4</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 CCAG… Monocytes        1 4       -0.0644      0.334     -0.0725     0.0560</span>
<span class="co">#&gt;  2 ATGA… Skeletal mu…     1 2       -0.178      -0.0167    -0.122      0.122 </span>
<span class="co">#&gt;  3 TATG… NK cells         1 6       -0.0604      0.0357     0.127      0.293 </span>
<span class="co">#&gt;  4 CATA… Monocytes        1 11      -0.0760      0.282     -0.121      0.0545</span>
<span class="co">#&gt;  5 GAGG… NK cells         1 1       -0.0758      0.0424     0.0977     0.259 </span>
<span class="co">#&gt;  6 AGCT… Fibroblasts      1 5       -0.117       0.0242    -0.0729     0.159 </span>
<span class="co">#&gt;  7 TGAT… B-cells          1 3       -0.146       0.0752    -0.143      0.0488</span>
<span class="co">#&gt;  8 ACGA… B-cells          1 3       -0.119       0.108     -0.127      0.0505</span>
<span class="co">#&gt;  9 CGGC… Monocytes        1 4       -0.00948     0.407     -0.0466     0.0803</span>
<span class="co">#&gt; 10 ATAG… Monocytes        1 4       -0.0561      0.367     -0.105      0.0589</span>
<span class="co">#&gt; # … with 2,889 more rows, and 4 more variables: corrected5 &lt;dbl&gt;, UMAP1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   UMAP2 &lt;dbl&gt;, UMAP3 &lt;dbl&gt;</span></code></pre></div>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/count.html">count</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">first.labels</span><span class="op">)</span>
<span class="co">#&gt; tidySingleCellExperiment says: A data frame is returned for independent data analysis.</span>
<span class="co">#&gt; # A tibble: 11 x 3</span>
<span class="co">#&gt;    label first.labels        n</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;           &lt;int&gt;</span>
<span class="co">#&gt;  1 1     NK cells          460</span>
<span class="co">#&gt;  2 10    B-cells            14</span>
<span class="co">#&gt;  3 11    Monocytes          26</span>
<span class="co">#&gt;  4 2     Skeletal muscle   795</span>
<span class="co">#&gt;  5 3     B-cells           367</span>
<span class="co">#&gt;  6 4     Monocytes         339</span>
<span class="co">#&gt;  7 5     Fibroblasts       486</span>
<span class="co">#&gt;  8 6     NK cells          276</span>
<span class="co">#&gt;  9 7     Smooth muscle      21</span>
<span class="co">#&gt; 10 8     Neutrophils        74</span>
<span class="co">#&gt; 11 9     Macrophages        41</span></code></pre></div>
</div>
</div>
<div id="key-points-1" class="section level2">
<h2 class="hasAnchor">
<a href="#key-points-1" class="anchor"></a>Key Points</h2>
<ul>
<li>Some basic steps of a single-cell RNA sequencing analysis are dimensionality reduction, cluster identification and cell type classification</li>
<li>
<code>tidySingleCellExperiment</code> is an invisible layer that operates on a <code>SingleCellExperiment</code> object and enables us to visualise and manipulate data as if it were a tidy data frame.</li>
<li>
<code>tidySingleCellExperiment</code> object is a <code>SingleCellExperiment object</code> so it can be used with any <code>SingleCellExperiment</code> compatible method</li>
</ul>
</div>
</div>
<div id="contributing" class="section level1">
<h1 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h1>
<p>If you want to suggest improvements for this workshop or ask questions, you can do so as described <a href="https://github.com/stemangiola/rladiestunis2021_tidytranscriptomics/blob/master/CONTRIBUTING.md">here</a>.</p>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>Record package and version information with <code>sessionInfo</code></p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] tidySingleCellExperiment_1.1.1 ggbeeswarm_0.6.0              </span>
<span class="co">#&gt;  [3] celldex_1.0.0                  EnsDb.Hsapiens.v86_2.99.0     </span>
<span class="co">#&gt;  [5] ensembldb_2.14.0               AnnotationFilter_1.14.0       </span>
<span class="co">#&gt;  [7] GenomicFeatures_1.42.3         AnnotationDbi_1.52.0          </span>
<span class="co">#&gt;  [9] scuttle_1.0.4                  SingleR_1.4.1                 </span>
<span class="co">#&gt; [11] batchelor_1.6.2                igraph_1.2.6                  </span>
<span class="co">#&gt; [13] scran_1.18.6                   scater_1.18.6                 </span>
<span class="co">#&gt; [15] SingleCellExperiment_1.12.0    tidySummarizedExperiment_1.1.4</span>
<span class="co">#&gt; [17] tidybulk_1.3.1                 tidyHeatmap_1.2.2             </span>
<span class="co">#&gt; [19] GGally_2.1.1                   ggrepel_0.9.1                 </span>
<span class="co">#&gt; [21] plotly_4.9.3                   purrr_0.3.4                   </span>
<span class="co">#&gt; [23] ggplot2_3.3.3                  stringr_1.4.0                 </span>
<span class="co">#&gt; [25] readr_1.4.0                    tidyr_1.1.3                   </span>
<span class="co">#&gt; [27] dplyr_1.0.5                    tibble_3.1.0                  </span>
<span class="co">#&gt; [29] airway_1.10.0                  SummarizedExperiment_1.20.0   </span>
<span class="co">#&gt; [31] Biobase_2.50.0                 GenomicRanges_1.42.0          </span>
<span class="co">#&gt; [33] GenomeInfoDb_1.26.7            IRanges_2.24.1                </span>
<span class="co">#&gt; [35] S4Vectors_0.28.1               BiocGenerics_0.36.0           </span>
<span class="co">#&gt; [37] MatrixGenerics_1.2.1           matrixStats_0.58.0            </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.1                               </span>
<span class="co">#&gt;   [2] tidyselect_1.1.0                         </span>
<span class="co">#&gt;   [3] RSQLite_2.2.6                            </span>
<span class="co">#&gt;   [4] htmlwidgets_1.5.3                        </span>
<span class="co">#&gt;   [5] grid_4.0.3                               </span>
<span class="co">#&gt;   [6] BiocParallel_1.24.1                      </span>
<span class="co">#&gt;   [7] munsell_0.5.0                            </span>
<span class="co">#&gt;   [8] ragg_1.1.2                               </span>
<span class="co">#&gt;   [9] preprocessCore_1.52.1                    </span>
<span class="co">#&gt;  [10] statmod_1.4.35                           </span>
<span class="co">#&gt;  [11] withr_2.4.1                              </span>
<span class="co">#&gt;  [12] colorspace_2.0-0                         </span>
<span class="co">#&gt;  [13] highr_0.8                                </span>
<span class="co">#&gt;  [14] knitr_1.31                               </span>
<span class="co">#&gt;  [15] rstudioapi_0.13                          </span>
<span class="co">#&gt;  [16] labeling_0.4.2                           </span>
<span class="co">#&gt;  [17] GenomeInfoDbData_1.2.4                   </span>
<span class="co">#&gt;  [18] bit64_4.0.5                              </span>
<span class="co">#&gt;  [19] farver_2.1.0                             </span>
<span class="co">#&gt;  [20] pheatmap_1.0.12                          </span>
<span class="co">#&gt;  [21] rprojroot_2.0.2                          </span>
<span class="co">#&gt;  [22] vctrs_0.3.7                              </span>
<span class="co">#&gt;  [23] generics_0.1.0                           </span>
<span class="co">#&gt;  [24] xfun_0.22                                </span>
<span class="co">#&gt;  [25] dittoSeq_1.2.5                           </span>
<span class="co">#&gt;  [26] BiocFileCache_1.14.0                     </span>
<span class="co">#&gt;  [27] R6_2.5.0                                 </span>
<span class="co">#&gt;  [28] clue_0.3-58                              </span>
<span class="co">#&gt;  [29] rsvd_1.0.3                               </span>
<span class="co">#&gt;  [30] locfit_1.5-9.4                           </span>
<span class="co">#&gt;  [31] bitops_1.0-6                             </span>
<span class="co">#&gt;  [32] cachem_1.0.4                             </span>
<span class="co">#&gt;  [33] reshape_0.8.8                            </span>
<span class="co">#&gt;  [34] DelayedArray_0.16.3                      </span>
<span class="co">#&gt;  [35] assertthat_0.2.1                         </span>
<span class="co">#&gt;  [36] promises_1.2.0.1                         </span>
<span class="co">#&gt;  [37] scales_1.1.1                             </span>
<span class="co">#&gt;  [38] beeswarm_0.3.1                           </span>
<span class="co">#&gt;  [39] debugme_1.1.0                            </span>
<span class="co">#&gt;  [40] gtable_0.3.0                             </span>
<span class="co">#&gt;  [41] beachmat_2.6.4                           </span>
<span class="co">#&gt;  [42] Cairo_1.5-12.2                           </span>
<span class="co">#&gt;  [43] rlang_0.4.10                             </span>
<span class="co">#&gt;  [44] genefilter_1.72.1                        </span>
<span class="co">#&gt;  [45] systemfonts_1.0.1                        </span>
<span class="co">#&gt;  [46] GlobalOptions_0.1.2                      </span>
<span class="co">#&gt;  [47] splines_4.0.3                            </span>
<span class="co">#&gt;  [48] rtracklayer_1.50.0                       </span>
<span class="co">#&gt;  [49] lazyeval_0.2.2                           </span>
<span class="co">#&gt;  [50] BiocManager_1.30.12                      </span>
<span class="co">#&gt;  [51] yaml_2.2.1                               </span>
<span class="co">#&gt;  [52] httpuv_1.5.5                             </span>
<span class="co">#&gt;  [53] tools_4.0.3                              </span>
<span class="co">#&gt;  [54] ellipsis_0.3.1                           </span>
<span class="co">#&gt;  [55] jquerylib_0.1.3                          </span>
<span class="co">#&gt;  [56] RColorBrewer_1.1-2                       </span>
<span class="co">#&gt;  [57] proxy_0.4-25                             </span>
<span class="co">#&gt;  [58] ggridges_0.5.3                           </span>
<span class="co">#&gt;  [59] Rcpp_1.0.6                               </span>
<span class="co">#&gt;  [60] plyr_1.8.6                               </span>
<span class="co">#&gt;  [61] sparseMatrixStats_1.2.1                  </span>
<span class="co">#&gt;  [62] progress_1.2.2                           </span>
<span class="co">#&gt;  [63] zlibbioc_1.36.0                          </span>
<span class="co">#&gt;  [64] RCurl_1.98-1.3                           </span>
<span class="co">#&gt;  [65] ps_1.6.0                                 </span>
<span class="co">#&gt;  [66] prettyunits_1.1.1                        </span>
<span class="co">#&gt;  [67] openssl_1.4.3                            </span>
<span class="co">#&gt;  [68] GetoptLong_1.0.5                         </span>
<span class="co">#&gt;  [69] viridis_0.5.1                            </span>
<span class="co">#&gt;  [70] cowplot_1.1.1                            </span>
<span class="co">#&gt;  [71] cluster_2.1.1                            </span>
<span class="co">#&gt;  [72] fs_1.5.0                                 </span>
<span class="co">#&gt;  [73] magrittr_2.0.1                           </span>
<span class="co">#&gt;  [74] RSpectra_0.16-0                          </span>
<span class="co">#&gt;  [75] data.table_1.14.0                        </span>
<span class="co">#&gt;  [76] ResidualMatrix_1.0.0                     </span>
<span class="co">#&gt;  [77] circlize_0.4.12                          </span>
<span class="co">#&gt;  [78] ProtGenerics_1.22.0                      </span>
<span class="co">#&gt;  [79] mime_0.10                                </span>
<span class="co">#&gt;  [80] hms_1.0.0                                </span>
<span class="co">#&gt;  [81] evaluate_0.14                            </span>
<span class="co">#&gt;  [82] xtable_1.8-4                             </span>
<span class="co">#&gt;  [83] XML_3.99-0.6                             </span>
<span class="co">#&gt;  [84] gridExtra_2.3                            </span>
<span class="co">#&gt;  [85] shape_1.4.5                              </span>
<span class="co">#&gt;  [86] compiler_4.0.3                           </span>
<span class="co">#&gt;  [87] biomaRt_2.46.3                           </span>
<span class="co">#&gt;  [88] crayon_1.4.1                             </span>
<span class="co">#&gt;  [89] htmltools_0.5.1.1                        </span>
<span class="co">#&gt;  [90] later_1.1.0.1                            </span>
<span class="co">#&gt;  [91] geneplotter_1.68.0                       </span>
<span class="co">#&gt;  [92] DBI_1.1.1                                </span>
<span class="co">#&gt;  [93] ExperimentHub_1.16.0                     </span>
<span class="co">#&gt;  [94] dbplyr_2.1.1                             </span>
<span class="co">#&gt;  [95] ComplexHeatmap_2.6.2                     </span>
<span class="co">#&gt;  [96] rappdirs_0.3.3                           </span>
<span class="co">#&gt;  [97] Matrix_1.3-2                             </span>
<span class="co">#&gt;  [98] cli_2.4.0                                </span>
<span class="co">#&gt;  [99] pkgconfig_2.0.3                          </span>
<span class="co">#&gt; [100] pkgdown_1.6.1                            </span>
<span class="co">#&gt; [101] GenomicAlignments_1.26.0                 </span>
<span class="co">#&gt; [102] xml2_1.3.2                               </span>
<span class="co">#&gt; [103] annotate_1.68.0                          </span>
<span class="co">#&gt; [104] vipor_0.4.5                              </span>
<span class="co">#&gt; [105] bslib_0.2.4                              </span>
<span class="co">#&gt; [106] dqrng_0.2.1                              </span>
<span class="co">#&gt; [107] XVector_0.30.0                           </span>
<span class="co">#&gt; [108] digest_0.6.27                            </span>
<span class="co">#&gt; [109] Biostrings_2.58.0                        </span>
<span class="co">#&gt; [110] rmarkdown_2.7                            </span>
<span class="co">#&gt; [111] uwot_0.1.10                              </span>
<span class="co">#&gt; [112] edgeR_3.32.1                             </span>
<span class="co">#&gt; [113] DelayedMatrixStats_1.12.3                </span>
<span class="co">#&gt; [114] curl_4.3                                 </span>
<span class="co">#&gt; [115] shiny_1.6.0                              </span>
<span class="co">#&gt; [116] Rsamtools_2.6.0                          </span>
<span class="co">#&gt; [117] rjson_0.2.20                             </span>
<span class="co">#&gt; [118] lifecycle_1.0.0                          </span>
<span class="co">#&gt; [119] jsonlite_1.7.2                           </span>
<span class="co">#&gt; [120] BiocNeighbors_1.8.2                      </span>
<span class="co">#&gt; [121] desc_1.3.0                               </span>
<span class="co">#&gt; [122] viridisLite_0.3.0                        </span>
<span class="co">#&gt; [123] askpass_1.1                              </span>
<span class="co">#&gt; [124] limma_3.46.0                             </span>
<span class="co">#&gt; [125] fansi_0.4.2                              </span>
<span class="co">#&gt; [126] pillar_1.5.1                             </span>
<span class="co">#&gt; [127] lattice_0.20-41                          </span>
<span class="co">#&gt; [128] rladiestunis2021tidytranscriptomics_0.8.0</span>
<span class="co">#&gt; [129] fastmap_1.1.0                            </span>
<span class="co">#&gt; [130] httr_1.4.2                               </span>
<span class="co">#&gt; [131] survival_3.2-10                          </span>
<span class="co">#&gt; [132] interactiveDisplayBase_1.28.0            </span>
<span class="co">#&gt; [133] glue_1.4.2                               </span>
<span class="co">#&gt; [134] FNN_1.1.3                                </span>
<span class="co">#&gt; [135] png_0.1-7                                </span>
<span class="co">#&gt; [136] BiocVersion_3.12.0                       </span>
<span class="co">#&gt; [137] bluster_1.0.0                            </span>
<span class="co">#&gt; [138] bit_4.0.4                                </span>
<span class="co">#&gt; [139] class_7.3-18                             </span>
<span class="co">#&gt; [140] stringi_1.5.3                            </span>
<span class="co">#&gt; [141] sass_0.3.1                               </span>
<span class="co">#&gt; [142] blob_1.2.1                               </span>
<span class="co">#&gt; [143] textshaping_0.3.3                        </span>
<span class="co">#&gt; [144] AnnotationHub_2.22.0                     </span>
<span class="co">#&gt; [145] DESeq2_1.30.1                            </span>
<span class="co">#&gt; [146] org.Hs.eg.db_3.12.0                      </span>
<span class="co">#&gt; [147] BiocSingular_1.6.0                       </span>
<span class="co">#&gt; [148] memoise_2.0.0                            </span>
<span class="co">#&gt; [149] irlba_2.3.3                              </span>
<span class="co">#&gt; [150] e1071_1.7-6</span></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-abbas2009deconvolution" class="csl-entry">
Abbas, Alexander R, Kristen Wolslegel, Dhaya Seshasayee, Zora Modrusan, and Hilary F Clark. 2009. <span>“Deconvolution of Blood Microarray Data Identifies Cellular Activation Patterns in Systemic Lupus Erythematosus.”</span> <em>PloS One</em> 4 (7): e6098.
</div>
<div id="ref-aran2019reference" class="csl-entry">
Aran, Dvir, Agnieszka P Looney, Leqian Liu, Esther Wu, Valerie Fong, Austin Hsu, Suzanna Chak, et al. 2019. <span>“Reference-Based Analysis of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic Macrophage.”</span> <em>Nature Immunology</em> 20 (2): 163–72.
</div>
<div id="ref-Cai2020" class="csl-entry">
Cai, Yi, Youchao Dai, Yejun Wang, Qianqing Yang, Jiubiao Guo, Cailing Wei, Weixin Chen, et al. 2020. <span>“Single-Cell Transcriptomics of Blood Reveals a Natural Killer Cell Subset Depletion in Tuberculosis.”</span> <em><span>EBioMedicine</span></em> 53 (March): 102686. <a href="https://doi.org/10.1016/j.ebiom.2020.102686">https://doi.org/10.1016/j.ebiom.2020.102686</a>.
</div>
<div id="ref-chen2016reads" class="csl-entry">
Chen, Yunshun, Aaron TL Lun, and Gordon K Smyth. 2016. <span>“From Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.”</span> <em>F1000Research</em> 5.
</div>
<div id="ref-Ding2020" class="csl-entry">
Ding, Jiarui, Xian Adiconis, Sean K. Simmons, Monika S. Kowalczyk, Cynthia C. Hession, Nemanja D. Marjanovic, Travis K. Hughes, et al. 2020. <span>“Systematic Comparison of Single-Cell and Single-Nucleus <span>RNA</span>-Sequencing Methods.”</span> <em>Nature Biotechnology</em> 38 (6): 737–46. <a href="https://doi.org/10.1038/s41587-020-0465-8">https://doi.org/10.1038/s41587-020-0465-8</a>.
</div>
<div id="ref-Freytag2018" class="csl-entry">
Freytag, Saskia, Luyi Tian, Ingrid Lönnstedt, Milica Ng, and Melanie Bahlo. 2018. <span>“Comparison of Clustering Tools in r for Medium-Sized 10x Genomics Single-Cell <span>RNA</span>-Sequencing Data.”</span> <em>F1000Research</em> 7 (December): 1297. <a href="https://doi.org/10.12688/f1000research.15809.2">https://doi.org/10.12688/f1000research.15809.2</a>.
</div>
<div id="ref-himes2014rna" class="csl-entry">
Himes, Blanca E, Xiaofeng Jiang, Peter Wagner, Ruoxi Hu, Qiyu Wang, Barbara Klanderman, Reid M Whitaker, et al. 2014. <span>“RNA-Seq Transcriptome Profiling Identifies Crispld2 as a Glucocorticoid Responsive Gene That Modulates Cytokine Function in Airway Smooth Muscle Cells.”</span> <em>PloS One</em> 9 (6): e99625.
</div>
<div id="ref-Karagiannis2020" class="csl-entry">
Karagiannis, Tanya T., John P. Cleary, Busra Gok, Andrew J. Henderson, Nicholas G. Martin, Masanao Yajima, Elliot C. Nelson, and Christine S. Cheng. 2020. <span>“Single Cell Transcriptomics Reveals Opioid Usage Evokes Widespread Suppression of Antiviral Gene Program.”</span> <em>Nature Communications</em> 11 (1). <a href="https://doi.org/10.1038/s41467-020-16159-y">https://doi.org/10.1038/s41467-020-16159-y</a>.
</div>
<div id="ref-law2016rna" class="csl-entry">
Law, Charity W, Monther Alhamdoosh, Shian Su, Xueyi Dong, Luyi Tian, Gordon K Smyth, and Matthew E Ritchie. 2016. <span>“RNA-Seq Analysis Is Easy as 1-2-3 with Limma, Glimma and edgeR.”</span> <em>F1000Research</em> 5.
</div>
<div id="ref-law2014voom" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. <span>“Voom: Precision Weights Unlock Linear Model Analysis Tools for RNA-Seq Read Counts.”</span> <em>Genome Biology</em> 15 (2): R29.
</div>
<div id="ref-love2014moderated" class="csl-entry">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15 (12): 550.
</div>
<div id="ref-mccarthy2012differential" class="csl-entry">
McCarthy, Davis J, Yunshun Chen, and Gordon K Smyth. 2012. <span>“Differential Expression Analysis of Multifactor RNA-Seq Experiments with Respect to Biological Variation.”</span> <em>Nucleic Acids Research</em> 40 (10): 4288–97.
</div>
<div id="ref-newman2015robust" class="csl-entry">
Newman, Aaron M, Chih Long Liu, Michael R Green, Andrew J Gentles, Weiguo Feng, Yue Xu, Chuong D Hoang, Maximilian Diehn, and Ash A Alizadeh. 2015. <span>“Robust Enumeration of Cell Subsets from Tissue Expression Profiles.”</span> <em>Nature Methods</em> 12 (5): 453–57.
</div>
<div id="ref-racle2017simultaneous" class="csl-entry">
Racle, Julien, Kaat de Jonge, Petra Baumgaertner, Daniel E Speiser, and David Gfeller. 2017. <span>“Simultaneous Enumeration of Cancer and Immune Cell Types from Bulk Tumor Gene Expression Data.”</span> <em>Elife</em> 6: e26476.
</div>
<div id="ref-robinson2010scaling" class="csl-entry">
Robinson, Mark D, and Alicia Oshlack. 2010. <span>“A Scaling Normalization Method for Differential Expression Analysis of RNA-Seq Data.”</span> <em>Genome Biology</em> 11 (3): 1–9.
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>&lt;maria.doyle at petermac.org&gt;<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>&lt;mangiola.s at wehi.edu.au&gt;<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
